gen.go

```diff
@@ -17,15 +18,40 @@ const MaxLength = 8192
 
 const ByteArrayMaxLen = 2 << 20
 
-const MaxLenTag = "maxlen"
-const NoUsrMaxLen = -1
+const (
+	MaxLenTag   = "maxlen"
+	NoUsrMaxLen = -1
+)
 
 var (
 	cidType      = reflect.TypeOf(cid.Cid{})
 	bigIntType   = reflect.TypeOf(big.Int{})
 	deferredType = reflect.TypeOf(Deferred{})
 )
 
+// CBORGeneric is a base interface for CBOR serializers that can be used as a generic type T such
+// that a serialization and deserialization methods will be called on any field of type T.
+//
+// Nullable fields are not handled by the cbor-gen generated code for generic types but should be
```

Comment:
reflect gives us no access to `T` or `*T`

I did try going down the path of allowing the user to specify this by choosing to supply a dummy value as a pointer or not, but this creates an inconsistency that we can't enforce nicely, they just have to be sure that if they generated it with a pointer then they don't supply a non-pointer type or they get a panic, and kind of vice versa. Without enforcing pointer or non-pointer in the generic itself you get unenforceable and undetectable (at generation time) edges. So at that point, it seems that the best punt is to make the user deal with it in their type.

Ideally cbor-gen should handle this itself for the root type so I can just use them as is as my generic param types - I should be able to say whether the type I'm generating has null handling at the top level, but it doesn't (doesn't consistently anyway) yet and this feels like a separate rabbit hole. e.g. when I do codegen for a `type CborInt int64` it would be nice to be able to say "this can also be null", but right now you can't.

Reply:
I'm talking about:

```go
type Foo[T CBORGeneric[T]] struct {
    Value *T
}
```

In this case, I'd expect cbor-gen to handle the pointer. However, given:

```go
type Foo[T CBORGeneric[T]] struct {
    Value T
}
```

I'd expect `T` to handle the pointer, if, e.g., `CBORGeneric[T]` was implemented with `*SomeType` as the receiver.

Reply:
Yeah, ok, you're right; I had a bug in my `toGenericTypeParam(string)` that was stopping this working so I thought it wasn't possible, but it turns out it is!

Check out `GenericStruct` and `SubGenericStruct` now, one without and one with pointer fields for their shared generic types. The tests then exercise that they both work properly, handling nulls in and out of both cases and it works fine.

The remaining awkwardness is in having `CBORGeneric` being different to `CBORSerializer` so having to add adapters to them just so we get a version of `UnmarshalCBOR` that returns an instance so we can handle both value and pointer types.

I did some further experimenting and if we accept a `New() V` on the `CBORGeneric` interface then we can use existing `CBORSerializer`s as generic types and support nullable fields too. The downside is that we get a bit tangled in pointer-to-pointer because of the need for a `UnmarshalCBOR` being on a pointer receiver. Check out the diff to make that work @ https://github.com/rvagg/cbor-gen/pull/1/files - if you look at `GenericStruct` and `SubGenericStruct` they are the same as the version here, but then look at the tests and you can see the need to do `&&` just to set values.